crypto/ec/ec_key.c

```diff
@@ -280,7 +281,17 @@ int ec_key_simple_generate_key(EC_KEY *eckey)
     eckey->pub_key = pub_key;
     priv_key = NULL;
     pub_key = NULL;
-
+#ifdef FIPS_MODE
+    {
+        OSSL_CALLBACK *cb = NULL;
+        void *cbarg = NULL;
+
+        /* TODO(3.0) - needs libctx */
+        OSSL_SELF_TEST_get_callback(NULL, &cb, &cbarg);
+        if (!SELF_TEST_keygen_pairwise_test_ecdsa(eckey, cb, cbarg))
+            goto err;
+    }
+#endif /* FIPS_MODE */
```

Comment:
Isn't there a function that takes a BN_GENCB somewhere?

Reply:
1) Its not even vaguely related to BN.
2) Its a self test - so it should be handled in a similiar manner to all the other self test events thru the same callback mechanism.

Reply:
To be blunt, it looks like hacks thrown in a little here and there rather than something done by design.  That's what I'm reacting against.  It's a little like having a global variable that gets used everywhere, and a few years ahead, you start wondering where the hell it's for, getting surprised by it being defined "over there", etc.

Reply:
re you talking about this code, or OPENSSL_CTX ? :)

Reply:
The requirement is that I want it to use the self test callback. Maybe if I shifted the code into the gen you would be happier..

Reply:
I'm talking about the SELF_TEST calls.  They reside in a module outside of libcrypto, while this is primarly libcrypto code.

True, it would make more sense to have these in the gen code proper, but most of all, it would make even more sense to pass down the callback and its extra argument to these functions than having them grab for it because the code happens to be part of the FIPS module.

Part of my concern is also that this kind of hack effectively excludes anyone else from using this kind of callback.  I have no doubt that other provider authors may want to include this kind of self test, especially if they create providers around something FIPS-able.

Reply:
This means the keygen function is going to need to pass in the callback into the keymanagement before it calls into the provider to do the actual keygen operation.. (I cant quite do that yet - but it will just be a moving the code about exercise). I dont think there is anything wrong with using the self test callback - I just need to move the code.

Reply:
> Part of my concern is also that this kind of hack effectively excludes anyone else from using this kind of callback. I have no doubt that other provider authors may want to include this kind of self test, especially if they create providers around something FIPS-able.

I don't understand your concern at all. Why is this excluding anyone?

Reply:
> Why is this excluding anyone?

...
...  I may have over-inflated the issue.

So, er, never mind.